### ISSUE TITLE
dd-trace module names conflicting with our app module names

### ISSUE THREAD

**Issue by tsebastian-sequoia** [DataDog/dd-trace-py#5601]
Image used: https://console.cloud.google.com/gcr/images/datadoghq/global/dd-lib-python-init@sha256:edb86a49bc10186875ff23a87353ea62abb16f7653ab55e8749979fb42092d48/details (dd-lib-python-init:v1.11.2)
dd-trace-py version used : v1.11.2

We have deployed an app in one of our K8 clusters and have deployed it with the dd-trace running as an init container. While our container was launching it was running into the following error:

```
Traceback (most recent call last):
  File "/usr/local/bin/uvicorn", line 8, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.8/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/usr/local/lib/python3.8/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/lib/python3.8/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/uvicorn/main.py", line 403, in main
    run(
  File "/usr/local/lib/python3.8/site-packages/uvicorn/main.py", line 568, in run
    server.run()
  File "/usr/local/lib/python3.8/site-packages/uvicorn/server.py", line 59, in run
    return asyncio.run(self.serve(sockets=sockets))
  File "/usr/local/lib/python3.8/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/usr/local/lib/python3.8/asyncio/base_events.py", line 616, in run_until_complete
    return future.result()
  File "/usr/local/lib/python3.8/site-packages/uvicorn/server.py", line 66, in serve
    config.load()
  File "/usr/local/lib/python3.8/site-packages/uvicorn/config.py", line 471, in load
    self.loaded_app = import_from_string(self.app)
  File "/usr/local/lib/python3.8/site-packages/uvicorn/importer.py", line 21, in import_from_string
    module = importlib.import_module(module_str)
  File "/usr/local/lib/python3.8/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "/usr/local/lib/python3.8/site-packages/ddtrace/internal/module.py", line 216, in _exec_module
    self.loader.exec_module(module)
  File "<frozen importlib._bootstrap_external>", line 843, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/app/src/main.py", line 11, in <module>
    from src.lib.redis import redis_cache
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "/usr/local/lib/python3.8/site-packages/ddtrace/internal/module.py", line 216, in _exec_module
    self.loader.exec_module(module)
  File "/app/src/lib/redis.py", line 3, in <module>
    import aioredis
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "/usr/local/lib/python3.8/site-packages/ddtrace/vendor/wrapt/importer.py", line 177, in _exec_module
    notify_module_loaded(module)
  File "/usr/local/lib/python3.8/site-packages/ddtrace/vendor/wrapt/decorators.py", line 470, in _synchronized
    return wrapped(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/ddtrace/vendor/wrapt/importer.py", line 136, in notify_module_loaded
    hook(module)
  File "/usr/local/lib/python3.8/site-packages/ddtrace/_monkey.py", line 151, in on_import
    imported_module = importlib.import_module(path)
  File "/usr/local/lib/python3.8/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "/usr/local/lib/python3.8/site-packages/ddtrace/internal/module.py", line 216, in _exec_module
    self.loader.exec_module(module)
  File "/usr/local/lib/python3.8/site-packages/ddtrace/contrib/aioredis/__init__.py", line 50, in <module>
    from .patch import patch
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "/usr/local/lib/python3.8/site-packages/ddtrace/internal/module.py", line 216, in _exec_module
    self.loader.exec_module(module)
  File "/usr/local/lib/python3.8/site-packages/ddtrace/contrib/aioredis/patch.py", line 22, in <module>
    from ..redis.util import _trace_redis_cmd
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "/usr/local/lib/python3.8/site-packages/ddtrace/internal/module.py", line 216, in _exec_module
    self.loader.exec_module(module)
  File "/usr/local/lib/python3.8/site-packages/ddtrace/contrib/redis/__init__.py", line 53, in <module>
    with require_modules(required_modules) as missing_modules:
  File "/usr/local/lib/python3.8/site-packages/ddtrace/internal/utils/importlib.py", line 20, in __init__
    import_module(module)
  File "/usr/local/lib/python3.8/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "/usr/local/lib/python3.8/site-packages/ddtrace/vendor/wrapt/importer.py", line 177, in _exec_module
    notify_module_loaded(module)
  File "/usr/local/lib/python3.8/site-packages/ddtrace/vendor/wrapt/decorators.py", line 470, in _synchronized
    return wrapped(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/ddtrace/vendor/wrapt/importer.py", line 136, in notify_module_loaded
    hook(module)
  File "/usr/local/lib/python3.8/site-packages/ddtrace/_monkey.py", line 157, in on_import
    imported_module.patch()
AttributeError: partially initialized module 'ddtrace.contrib.redis' has no attribute 'patch' (most likely due to a circular import)

```
This looks like it is being caused due to the ddtrace module named 'redis' conflicting with one of our app module named 'redis'.

Steps to reproduce:
1. Create a python app with a local module named 'redis' (or same name as any other module under ddtrace).
2. Configure ddtrace to launch the python app and observe the stacktrace for the same.

Since ddttrace  is having common names for its modules it is possible that in future we could be running into similar issues with other modules. The fix for this is to update the user's app module names with something else other than the ones used by ddtrace which is not practical since it would require ddtrace users to rename their app folders for ddtrace to work.

Please look into a fix for this issue from ddtrace end.

**Comment by Yun-Kim** [comment]
Hi @tsebastian-sequoia, this issue should be resolved by fix #5608. 
Addressing your second point about ddtrace module names conflicting with your app module names, from what I can tell this isn't the underlying problem, the problem was cross-integration imports as @P403n1x87 mentioned. ddtrace module names conflicting with your app module names would only be a problem if you were somehow using both the `redis` package and using ddtrace's `redis` integration as well as having your own `redis.py` file.

**Comment by qu3vipon** [comment]
Is there any workaround for this issue before v1.14.0 is released?

**Comment by ZStriker19** [comment]
> 

Hi @qu3vipon we've backported the fix so it'll be out in 1.12.3 as well, which should be released shortly.
